README.md
# MultipleChoiceAnswerBot
A Python program that prompts user for a question and multiple choice answers. Retrieve data from Google using Rapid API, and compute the most correct answer out of the following choices.

![Screenshot](https://github.com/AlvinKW/MultipleChoiceAnswerBot/blob/main/example.gif)

## Getting Started

### Prerequisites
Things you need to install:

```
nltk
Pillow
pytesseract
regex
requests
tesseract
```

### Installing


1. Clone the repository
```
git clone https://github.com/AlvinKW/MultipleChoiceAnswerBot.git
```
2. cd to directory where requirements.txt is located
```
Run pip install -r requirements.txt
```
3. Install tesseract
```
Windows: Install tesseract from https://github.com/UB-Mannheim/tesseract/wiki
         Create a folder Tesseract-OCR where requirements.txt is located
         Install exe file inside Tesseract-OCR
         Uncomment tess.pytesseract.tesseract_cmd = 'Tesseract-OCR/tesseract.exe' in computing.py file
MacOS: brew install tesseract
```
4. Get API Key
```
https://rapidapi.com/apigeek/api/google-search3/endpoints
```

### Running the application
cd to where main.py located and in terminal use commands:
```
python3 main.py
```


## Author

Alvin Kwan


project/history.txt
Question: 'Who is the first president of the United States?'
Multiple Choices: 
	a) george washington  = 64.13%
	b) bill gates  = 0.00%
	c) donald trump  = 21.74%
	d) barack obama  = 13.04%
	e) abraham lincoln  = 1.09%

The most correct answer would be: a) george washington 
-------------------------------------------------------
Question: 'who is the current president?'
Multiple Choices: 
	barack obama  = 11.25%
	donald trump  = 78.75%
	george washington  = 10.00%
	alvin kwan  = 0.00%

The most correct answer would be: donald trump 
-------------------------------------------------------
Question: 'What is the capital of Maine?'
Multiple Choices: 
	boston  = 0.00%
	austin  = 0.00%
	augusta  = 100.00%

The most correct answer would be: augusta 
-------------------------------------------------------
Question: 'How many bones are in the human body?'
Multiple Choices: 
	a) 206  = 100.00%
	b) 205  = 0.00%
	c) 201  = 0.00%
	d) 209  = 0.00%

The most correct answer would be: a) 206 
-------------------------------------------------------
Question: 'what is the capital of massachusetts?'
Multiple Choices: 
	boston  = 100.00%
	sacramento  = 0.00%
	autson  = 0.00%
	augustus  = 0.00%

The most correct answer would be: boston 
-------------------------------------------------------
Question: 'How many bones are in the human body?'
Multiple Choices: 
	a) 206  = 100.00%
	b) 205  = 0.00%
	c) 201  = 0.00%
	d) 209  = 0.00%

The most correct answer would be: a) 206 
-------------------------------------------------------
Question: 'How many bones are in the human body?'
Multiple Choices: 
	a) 206  = 100.00%
	b) 205  = 0.00%
	c) 201  = 0.00%
	d) 209  = 0.00%

The most correct answer would be: a) 206 
-------------------------------------------------------
Question: 'which company sells iphones?'
Multiple Choices: 
	apple  = 100.00%
	tesla  = 0.00%
	microsoft  = 0.00%
	the  = 0.00%

The most correct answer would be: apple 
-------------------------------------------------------
Question: 'What is the name for the Jewish New Year?'
Multiple Choices: 
	a) hanukkah  = 0.00%
	b) yom kippur  = 15.48%
	c) kwanza  = 0.00%
	d) rosh hashanah  = 84.52%

The most correct answer would be: d) rosh hashanah 
-------------------------------------------------------
Question: 'What is the capital of Massachusetts?'
Multiple Choices: 
	a) washington  = 1.85%
	b) boston  = 96.30%
	c) sacramento  = 0.00%
	d) little rock  = 0.00%
	e) phoenix  = 1.85%

The most correct answer would be: b) boston 
-------------------------------------------------------
Question: 'What city is boston university located?'
Multiple Choices: 
	boston  = 99.47%
	newton  = 0.53%
	dedham  = 0.00%
	natick  = 0.00%

The most correct answer would be: boston 
-------------------------------------------------------
Question: 'what color is the sky?'
Multiple Choices: 
	red  = 13.76%
	yellow  = 6.42%
	blue  = 72.48%
	white  = 7.34%

The most correct answer would be: blue 
-------------------------------------------------------
Question: 'What is the capital of Germany?'
Multiple Choices: 
	a) berlin  = 91.38%
	b) munich  = 3.45%
	c) hamburg  = 5.17%
	d) boston  = 0.00%

The most correct answer would be: a) berlin 
-------------------------------------------------------
Question: 'Fe is the chemical symbol for?'
Multiple Choices: 
	zinc  = 1.55%
	hydrogen  = 1.55%
	fluorine  = 1.55%
	iron  = 95.35%

The most correct answer would be: iron 
-------------------------------------------------------
Question: 'Which country gifted the statue of liberty to the U.S.?'
Multiple Choices: 
	germany  = 0.00%
	china  = 2.27%
	france  = 95.45%
	italy  = 2.27%

The most correct answer would be: france 
-------------------------------------------------------
Question: 'Who is the first president of the United States?'
Multiple Choices: 
	a) george washington  = 79.25%
	b) bill gates  = 0.00%
	c) donald trump  = 11.32%
	d) barack obama  = 8.49%
	e) abraham lincoln  = 0.94%

The most correct answer would be: a) george washington 
-------------------------------------------------------
Question: 'What is the capital of Germany?'
Multiple Choices: 
	a) berlin  = 95.74%
	b) munich  = 4.26%
	c) hamburg  = 0.00%
	d) boston  = 0.00%

The most correct answer would be: a) berlin 
-------------------------------------------------------
Question: 'which one came first the chicken or the egg?'
Multiple Choices: 
	chicken  = 50.29%
	egg  = 49.71%

The most correct answer would be: chicken 
-------------------------------------------------------
Question: 'AO. Trivia Question: Which country gifted the Statue of Liberty to the U.S.?'
Multiple Choices: 
	a) germany  = 4.35%
	b) china  = 4.35%
	c) france  = 91.30%
	d) italy  = 0.00%

The most correct answer would be: c) france 
-------------------------------------------------------
Question: 'Who is the first president of the United States?'
Multiple Choices: 
	a) george washington  = 84.76%
	b) bill gates  = 0.00%
	c) donald trump  = 9.52%
	d) barack obama  = 4.76%
	e) abraham lincoln  = 0.95%

The most correct answer would be: a) george washington 
-------------------------------------------------------
Question: 'Where is Boston University located?'
Multiple Choices: 
	boston  = 100.00%
	newton  = 0.00%
	natick  = 0.00%
	burlington  = 0.00%

The most correct answer would be: boston 
-------------------------------------------------------
Question: 'Who is the first president of the United States?'
Multiple Choices: 
	a) george washington  = 66.67%
	b) bill gates  = 4.30%
	c) donald trump  = 23.66%
	d) barack obama  = 3.23%
	e) abraham lincoln  = 2.15%

The most correct answer would be: a) george washington 
-------------------------------------------------------
Question: 'Where is boston university located?'
Multiple Choices: 
	boston  = 100.00%
	newton  = 0.00%
	natick  = 0.00%
	burlington  = 0.00%

The most correct answer would be: boston 
-------------------------------------------------------


project/test_class_methods.py
"""
Alvin Kwan
Class: CS 521 - Fall 2
Date: December 17, 2020
CS521 Project
Unit Test
The project is designed to answer multiple choice questions through gathering
data from Google and basing off of the frequencies of the choices.
"""
import unittest
from computing import Search

class TestSearch(unittest.TestCase):
    '''
    Class for unit testing modules from computing.py
    '''
    def test_calculate_score(self):
        '''
        Function to test calculate_score() single words
        '''
        test = Search('', 'apple orange grape', ['apple', 'orange', 'grape'])
        test_score = {'apple':5, 'orange':10, 'grape':20}
        self.assertEqual(test.calculate_score(test_score),
                         {'apple':5, 'orange':10, 'grape':20})

    def test_calculate_score_multiwords(self):
        '''
        Function to test calculate_score() multiwords
        '''
        test = Search('', 'Elon Musk Jeff Bezos  Tim Apple',
                      ['Elon Musk', 'Jeff Bezos', 'Tim Apple'])
        test_score = {'Elon': 10, 'Musk': 20, 'Jeff': 30, 'Bezos': 40, 'Tim': 50,
                      'Apple': 60}
        self.assertEqual(test.calculate_score(test_score),
                         {'Elon Musk': 30, 'Jeff Bezos': 70, 'Tim Apple': 110})

    def test_print_choice(self):
        '''
        Function to test print_choice
        '''
        test = Search('', 'apple orange grape', ['apple', 'orange', 'grape'])
        self.assertEqual(test.print_choice(),
                         ('Multiple Choices: \n', {'apple': 0, 'orange': 0, 'grape': 0}))

    def test_print_result(self):
        '''
        Function to test print_result()
        '''
        test = Search('', 'apple orange grape', ['apple', 'orange', 'grape'])
        dash = '\n' + '-'*55
        output = "Multiple Choices: \n"
        msg = '\nError: answer cannot be determined'
        result = output + msg + dash
        self.assertEqual(test.print_result(), result)

    def test_repr(self):
        '''
        Function to test __repr__()
        '''
        test = Search('This is a question', 'apple orange grape',
                      ['apple', 'orange', 'grape'])
        self.assertEqual(test.__repr__(), "\'This is a question?\'")

if __name__ == '__main__':
    unittest.main()
    

project/main.py
"""
Alvin Kwan
Class: CS 521 - Fall 2
Date: December 17, 2020
CS521 Project
Main File
The project is designed to answer multiple choice questions through gathering
data from Google and basing off of the frequencies of the choices.
"""
import os.path
from computing import UserImage, UserInput, Search

def main():
    '''Main method. Prompts user for input based on the following choices:
    Upload file, input manual question, exit program.
    Arguments:
    None

    Returns:
    None
    '''
    while True:
        # Print dialogue to user
        print('\n' + "#"*55 + '\n')
        print('Enter a choice: \n')
        print('1. Upload multiple choice question through image file')
        print('2. Enter multiple choice question manually')
        print('3. Exit program\n')
        user_input = input()
        # Import image from file and compute result of multiple choice question
        if user_input == '1':
            user_img = input('Enter image file located on same directory as main.py: ')
            if 'png' in user_img and os.path.exists(user_img):
                option_one(user_img)
            else:
                print('Error: File not found')
        # Get user's manual question and compute result
        elif user_input == '2':
            option_two()
        # End program
        elif user_input == '3':
            print("Program has ended.")
            break
        else:
            print("Invalid input.")

def option_one(user_img):
    '''Get image from user and convert to String and import to output.txt
    From output.txt, parse question and muliple choices and return best answer
    Arguments:
    (str) image name

    Returns:
    None
    '''
    # Convert image to String and input to output.txt
    user_image = UserImage(user_img)
    user_image.convert_images()
    # Parse question from output.txt
    valid, msg = get_question()
    # Check if question is valid
    if valid:
        choices_string = ''
        choices_list = get_choices()
        question = msg
        # Create choice string from choice list
        for item in choices_list:
            choices_string += item + '\n'
        # Search object to search question
        new_search = Search(question, choices_string, choices_list)
        print('-'*55)
        # Result
        search_result = new_search.print_result()
        print(search_result)
        # Prompt user if he/she want to save result to history.txt
        while True:
            user_save = input("Do you want to save result to history.txt: (Y/N)? ")
            if user_save.lower() == 'y':
                save_to_history(search_result)
                break
            elif user_save.lower() == 'n':
                break
            else:
                print("Error: invalid input")
    else:
        print(msg)

def option_two():
    '''Prompt user to enter multiple choice questions and chain of possible
    answers. Import string to output.txt. From output.txt, parse question
    and muliple choices and return best answer.
    Arguments:
    None

    Returns:
    None
    '''
    # Prompt user for multiple choice question
    user_question = input("Enter multiple choice question: ")
    print('\n')
    choices_string = ''
    # Check if question is valid
    user_input_ob = UserInput(user_question, choices_string)
    user_input_ob.input_to_file()
    valid, msg = get_question()
    if valid:
        # Prompt user for multiple choice answers, delimited by new line
        user_choices = []
        print("Enter multiple choices (DO NOT ENTER NUMBERS): ")
        print("After each individual choice, hit ENTER.")
        print("Enter '0' to state you are finished adding choices.")
        # Keep entering answers while condition is TRUE
        while True:
            user_input = input()
            # Make sure user does not enter a number other than 0
            if user_input != '0' and user_input.isdigit():
                print('{} will not be considered. Please do not enter numbers.'.format(user_input))
            # '0' means no more answers to input
            elif user_input == '0':
                choices_string = ''
                # Create choice string from choice list
                for item in user_choices:
                    choices_string += item + '\n'
                user_input_second = UserInput(user_question, choices_string)
                user_input_second.input_to_file()
                choices_string = ''
                choices_list = get_choices()
                question = msg
                for item in choices_list:
                    choices_string += item
                # Search object to search question
                new_search = Search(question, choices_string, choices_list)
                print('-'*55)
                # Result
                search_result = new_search.print_result()
                print(search_result)
                # Prompt user if he/she want to save result to history.txt
                while True:
                    user_save = input("Do you want to save result to history.txt: (Y/N)? ")
                    if user_save.lower() == 'y':
                        save_to_history(search_result)
                        break
                    elif user_save.lower() == 'n':
                        break
                    else:
                        print("Error: invalid input")
                break
            else:
                user_choices.append(user_input)
    else:
        print(msg)

def save_to_history(answer):
    '''Save string to history.txt file by appending.
    Arguments:
    (String) result string

    Returns:
    None
    '''
    file = open('history.txt', 'a')
    file.writelines(answer)
    file.close()

def get_question():
    '''Function to retrieve question from output.txt
    Arguments:
    None

    Returns:
    (bool) question is valid or not
    (str) message for error otherwise return question
    '''
    # Return none and false if not a question
    # otherwise return question
    valid = True
    error_msg = None
    file = open('output.txt', 'r')
    question = ''
    for line in file:
        sentence = line.split()
        for word in sentence:
            question += word + ' '
    temp = question
    # Separate question from other text
    sep = '?'
    question = question.split(sep, 1)[0]
    if sep not in temp or len(question) < 5:
        valid = False
        error_msg = 'Error: invalid question (make sure to end question with \'?\')'
        return valid, error_msg
    else:
        return valid, question

def get_choices():
    '''Function to retrieve choices from output.txt
    Arguments:
    None

    Returns:
    None
    '''
    choice_list = []
    final_list = []
    file = open('output.txt', 'r')
    # Iterate through file for choices
    for line in file:
        choice_list.append(line.replace('\n', ''))
    for i in choice_list:
        # Remove question and only input choices
        if '?' in i:
            choice_list = choice_list[choice_list.index(i) + 1:len(choice_list)]
    # Iterate through choices and only save real word choices
    for i in choice_list:
        temp = ''
        choice = i.split()
        for word in choice:
            temp += word + ' '
        temp = temp.lower()
        final_list.append(temp)
    # Remove empty string
    final_list = list(filter(None, final_list))
    return final_list
# Main
if __name__ == '__main__':
    main()


project/computing.py
"""
Alvin Kwan
Class: CS 521 - Fall 2
Date: December 17, 2020
CS521 Project
Modules File
API key retrieved from: https://rapidapi.com/apigeek/api/google-search3/
tutorials/using-python-to-call-google-search-api
The project is designed to answer multiple choice questions through gathering
data from Google and basing off of the frequencies of the choices.
"""
import re
import requests
import pytesseract as tess
import nltk
from PIL import Image
from nltk.corpus import stopwords
nltk.download('stopwords')
###########################################################################
# If NLTK SSL error occurs uncomment
# import nltk
# import ssl

# try:
#     _create_unverified_https_context = ssl._create_unverified_context
# except AttributeError:
#     pass
# else:
#     ssl._create_default_https_context = _create_unverified_https_context

# nltk.download()
############################################################################

# Uncomment if you are using WINDOWS
# tess.pytesseract.tesseract_cmd = 'Tesseract-OCR/tesseract.exe'
class UserImage:
    '''
    Class for retrieving user image and converting to text
    ...

    Attributes
    ----------
    None

    Methods
    ----------
    convert_images():
        Convert user image to text and write into a text document
    '''

    def __init__(self, image):
        '''
        Constructor for UserImage
        '''
        self.image = image

    def convert_images(self):
        '''Converts user input image to String and import to a text document
        Arguments:
        None

        Returns:
        None
        '''
        # Check if file exist
        try:
            img = Image.open(self.image)
        except IOError:
            print("Error: {} file does not exist".format(self.image))
        else:
            text = tess.image_to_string(img)
            file = open('output.txt', 'w')
            file.writelines('')
            file = open('output.txt', 'a')
            file.writelines(text)
            file.close()

class UserInput:
    '''
    Class for writing user questiona and multiple choice to text document
    ...

    Attributes
    ----------
    None

    Methods
    ----------
    input_to_file():
        Write question and multiple choice answers to text document
    '''
    def __init__(self, question, choice):
        '''
        Constructor for UserInput object
        '''
        self.question = question
        self.choice = choice

    def input_to_file(self):
        '''Writes question and multiple choice to text file
        Arguments:
        None

        Returns:
        None
        '''
        file = open('output.txt', 'w')
        file.writelines('')
        file = open('output.txt', 'a')
        file.writelines(self.question)
        file.writelines('\n')
        file.writelines(self.choice)
        file.close()

class Search:
    '''
    A class that represent a search object
    ...

    Attributes
    ----------
    None

    Methods
    ----------
    __repr__():
        Returns a printable representation of the question
    __parse_search():
        Returns a dictionary of multiple choices and their frequency
    calculate_score():
        Returns a dictionary of multiple choice computed scores
    print_choice():
        Returns multiple choice and percentages, and dictionary from
        parseSearch()
    print_result():
        Prints result of the best choice based on calculations of text from google
    '''
    def __init__(self, question, choice, choice_list):
        '''
        Constructor for Search object
        '''
        self.__question = question
        self.__choice = choice
        self.__choice_list = choice_list

    def __repr__(self):
        '''Returns printable representation of object question
        Argument:
        None

        Returns:
        (String) printable representation of question
        '''
        return repr(self.__question + '?')

    def __parse_search(self):
        '''Check for frequency of words from question searched using Google Search API.
        Argument:
        None

        Returns:
        (dict) dictionary of words choices and their frequency
        '''
        word_dict = {}
        word = self.__choice.split()
        # Use Rapid Google API to query the text data
        url = "https://google-search3.p.rapidapi.com/api/v1/search/q={}&num=100".format(
            self.__question
        )
        # API Key
        headers = {
            'x-rapidapi-key': "b1fc6ac51dmsh35a6d981032b6b4p1c6a98jsnf36bb6f64b27",
            'x-rapidapi-host': "google-search3.p.rapidapi.com"
            }

        response = requests.request("GET", url, headers=headers)
        text = response.text
        text = text.lower()
        # Use regex to remove all punctuation
        text = re.sub(r'[^\w\s]', '', text)
        list_text = text.split()
        # Remove all stopwords
        list_text = [w for w in list_text if w not in stopwords.words('english')]
        # Initialize word dictionary to keep track of each word count
        # shown up in the text
        for item in word:
            word_dict[item] = 0
        # Iterate through the online text data and see if it matches
        # a word from choice list, increment by 1 if in list
        for item in list_text:
            if item in word:
                if item not in word_dict:
                    word_dict[item] = 1
                else:
                    word_dict[item] += 1
        return word_dict

    def calculate_score(self, words):
        '''Calculate the score of each multiple choice answers.
        Arguments:
        (dict) dictionary of words with their frequency

        Returns:
        (dict) dictionary of choices and computed scores
        '''
        choice_dict = {}
        # Initialize choice dictionary to keep track of frequency of each choice
        for item in self.__choice_list:
            choice_dict[item] = 0

        for key, value in words.items():
            for word in self.__choice_list:
                if key in word:
                    choice_dict[word] += value
        return choice_dict

    def print_choice(self):
        '''Helper function to return choices and their frequencies and dictionary used.
        Argumens:
        None

        Returns:
        (String) Multiple choice answers and their percentage
        (dict) Dictionary of multiple choicees and their frequencies
        '''
        title = "Multiple Choices: \n"
        output = title
        total = 0
        # Dictionary for holding words and their frequencies
        word_dict = self.__parse_search()
        for key, val in self.calculate_score(word_dict).items():
            total += val
        # Catch error if total equals zero
        try:
            if total != 0:
                # Iterate through dictionary and the sum of multiple choices
                for key, val in self.calculate_score(word_dict).items():
                    temp = (val/total)*100
                    output += '\t{} = {:,.2f}%\n'.format(key, temp)
            return output, word_dict
        except ZeroDivisionError as e:
            print("Error: {}".format(e))

    def print_result(self):
        '''Print function for result output.
        Arguments:
        None

        Returns:
        (String) Result text
        '''
        # Print the question object
        question_string = 'Question: ' + repr(Search(
            self.__question, self.__choice, self.__choice_list)) + '\n'
        print('Please wait...Processing...Computing Answer...\n')
        output, word_dict = self.print_choice()
        max_val = 0
        temp = 0
        # Compute the weight sum for each mutiple choice answers
        calculate_dict = self.calculate_score(word_dict)
        num_list = []
        # Initizalize list
        duplicate_list = [0]
        # Get answer with the highest value
        for key, val in calculate_dict.items():
            temp = val
            if max_val < temp:
                max_val = temp
                max_key = key
            if val not in num_list:
                num_list.append(val)
            else:
                duplicate_list.append(val)
        # Print conditions
        # Print error if answers not determined
        if max_val == 0:
            dash = '\n' + '-'*55
            msg = '\nError: answer cannot be determined'
            result = output + msg + dash
            return result
        # Print duplicates for multiple answers
        elif max_val == max(duplicate_list):
            duplicate_choice = []
            for key, val in calculate_dict.items():
                if max_val == val:
                    duplicate_choice.append(key)
            dash = '\n' + '-'*55
            msg = '\nThe most correct answers would be: {}'.format(duplicate_choice)
            result = output + msg + dash
            return result
        # Print answer
        else:
            dash = '\n' + '-'*55 + '\n'
            msg = '\nThe most correct answer would be: {}'.format(max_key)
            result = question_string + output + msg + dash
            return result
